Common/SemanticAnalyzer/src/SemanticAnalyzer.h

```diff
@@ -0,0 +1,256 @@
+    std::string id2;
+    std::vector<Property> properties;
+};
+
+struct Note
+{
+    std::string id;
+    std::vector<Property> properties;
+};
+
+struct Graph
+{
+    std::string id;
+    std::vector<Property> properties;
+    std::vector<ObjectDecl> objects;
+    std::vector<Relation> relations;
+};
+
+struct DotCloud
+{
+    std::string id;
+    std::vector<Property> externalProperties;
+    std::vector<Property> internalProperties;
+};
+
+struct ConstraintInfo
+{
+    std::string type;
+    std::optional<double> min;
+    std::optional<double> max;
+};
+
+
+inline static const std::vector<std::string> COMMON_PROPERTIES =
+{
+    PROP_COLOR,
+    PROP_TEXT,
+    PROP_BORDER,
+    PROP_X,
+    PROP_Y,
+    PROP_SIZE_TEXT
+};
+
+inline static const std::map<std::string, std::vector<std::string>> SHAPE_SPECIFIC_PROPERTIES =
+{
+    {SHAPE_CIRCLE,    {PROP_RADIUS}},
+    {SHAPE_RECTANGLE, {PROP_SIZE_A, PROP_SIZE_B}},
+    {SHAPE_DIAMOND,   {PROP_SIZE_A, PROP_SIZE_B, PROP_ANGLE}},
+    {DOT_CLOUD,       {PROP_GRID}}
+};
+
+inline static const std::map<std::string, ConstraintInfo> PROPERTY_CONSTRAINTS =
+{
+    {PROP_RADIUS,    {TYPE_NUMBER, 0.0, std::nullopt}},
+    {PROP_SIZE_A,    {TYPE_NUMBER, 0.0, std::nullopt}},
+    {PROP_SIZE_B,    {TYPE_NUMBER, 0.0, std::nullopt}},
+    {PROP_ANGLE,     {TYPE_NUMBER, 0.0, 360.0}},
+    {PROP_COLOR,     {TYPE_STRING, std::nullopt, std::nullopt}},
+    {PROP_TEXT,      {TYPE_STRING, std::nullopt, std::nullopt}},
+    {PROP_BORDER,    {TYPE_NUMBER, 0.0, std::nullopt}},
+    {PROP_X,         {TYPE_NUMBER, std::nullopt, std::nullopt}},
+    {PROP_Y,         {TYPE_NUMBER, std::nullopt, std::nullopt}},
+    {PROP_SIZE_TEXT, {TYPE_NUMBER, 0.0, std::nullopt}},
+    {PROP_GRID,      {TYPE_BOOLEAN, std::nullopt, std::nullopt}}
+};
+
+inline static const std::vector<std::string> ALLOWED_COLORS =
+{
+    RED, GREEN, BLUE, BLACK, WHITE, YELLOW, PURPLE
+};
+
+
+class SemanticError: public std::runtime_error
+{
+public:
+    SemanticError(const std::string& message, int statementNumber)
+        : std::runtime_error("Ошибка: " + message + " Statement " + std::to_string(statementNumber)){}
+};
+
+
+class SEMANTICANALYZER_API SemanticAnalyzer
+{
+public:
+
+    static SemanticAnalyzer& getInstance()
+    {
+        static SemanticAnalyzer instance;
+        return instance;
+    }
+
+    /*!
+     * \brief Точка входа в семантический анализатор.
+     * \param programTree - набор Statement
+     * \param statementNumber - номер Statement для логирования(временное решение)
+     */
+    void semanticAnalysis(const std::vector<std::pair<std::string, std::any>>& programTree, int statementNumber = 1);
+
+    /*!
+     * \brief Сброс состояния анализатора.
+     */
+    void reset() { scopeStack.clear(); }
+
+    /*!
+     * \brief  Функция возвращает все объявленные глобально объекты.
+     */
+    const std::set<std::string>& getDeclaredObjects() const { return scopeStack.empty() ? emptySet : scopeStack.back(); }
```

Comment:
[minor] Тут дело вкуса, но можно не указывать emptySet, а просто указать
```suggestion
    const std::set<std::string>& getDeclaredObjects() const { return scopeStack.empty() ? {} : scopeStack.back(); }
```
Но тут сам смотри, что так, что так правильно

Reply:
у меня твой вариант вообще не компилируется:(

Reply:
Блин, так я тебе фигню сказал, кажется, что у нас вообще битая ссылка должна быть в таком случае. Лучше оставь свой